File: FetchApp/src/middlewares/authorize-jwt.js
const jwt = require('jsonwebtoken');
const config = require('../config')

function authorizeJWT(req, res, next) {
  const authHeader = req.headers.authorization;
  const token = authHeader && authHeader.split(' ')[1];

  if (token == null) {
      return res.status(401).json({ message: 'Unauthorized' });
  }

//   console.log('===================', token, config.jwt.JWTSecret)

  jwt.verify(token, config.jwt.JWTSecret, function(err, decoded) {
      if (err) {
          return res.status(401).json({ message: 'Unauthorized' });
      }
    //   console.log("passssssss")

      req.decoded = decoded;

      next();
  });
}

module.exports = authorizeJWT;

File: AuthApp/middlewares/authorize_jwt.go
package middlewares

import (
	"AuthApp/config"
	"AuthApp/httperror"
	"AuthApp/models"
	"encoding/json"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/golang-jwt/jwt/v4"
)

func validateToken(encodedToken string) (*jwt.Token, error) {
	return jwt.Parse(encodedToken, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, httperror.Unauthorized()
		}

		return config.Config.JWTSecret, nil
	})
}

func AuthorizeJWT(c *gin.Context) {
	authHeader := c.GetHeader("Authorization")

	s := strings.Split(authHeader, "Bearer ")
	unauthorizedErr := httperror.Unauthorized()
	if len(s) < 2 {
		c.AbortWithStatusJSON(unauthorizedErr.StatusCode, unauthorizedErr)
		return
	}
	encodeToken := s[1]

	token, err := validateToken(encodeToken)
	if err != nil || !token.Valid {
		c.AbortWithStatusJSON(unauthorizedErr.StatusCode, unauthorizedErr)
		return
	}
	claims, ok := token.Claims.(jwt.MapClaims)
	if !ok {
		c.AbortWithStatusJSON(unauthorizedErr.StatusCode, unauthorizedErr)
		return
	}

	//bind
	userJson, err := json.Marshal(claims["user"])
	var user models.User
	err = json.Unmarshal(userJson, &user)
	if err != nil {
		c.AbortWithStatusJSON(unauthorizedErr.StatusCode, unauthorizedErr)
		return
	}
	c.Set("user", user)
}
